cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +193,198 @@ constexpr bool is_hex_digit(char ch)
+        int const char_idx = char_idx_start + warp_lane;
+        int8_t const is_ichar_escape_char =
+          (char_idx < string_length_block && is_escape_char(in_chars_shared + char_idx)) ? 1 : 0;
+
+        // Warp-wise reduction to calculate the number of escape characters.
+        // All threads in the warp participate in the reduction, even if `char_idx` is beyond
+        // `string_length_block`.
+        int8_t const total_escape_char =
+          cub::WarpReduce<int8_t>(cub_storage[local_warp_id]).Sum(is_ichar_escape_char);
+
+        if (warp_lane == 0) { escape_char_count += total_escape_char; }
+
+        __syncwarp();
+      }
+    }
+    // URL decoding replaces 3 bytes with 1 for each escape character.
+    if (warp_lane == 0) { out_counts[row_idx] = string_length - escape_char_count * 2; }
   }
-};
+}
 
 /**
- * @brief Functor to update the string column offsets.
+ * @brief Decode and copy from the input char column to the output char column.
+ *
+ * @tparam num_warps_per_threadblock Number of warps in a threadblock. This template argument must
+ * match the launch configuration, i.e. the kernel must be launched with
+ * `num_warps_per_threadblock * cudf::detail::warp_size` threads per threadblock.
+ * @tparam char_block_size Number of characters which will be loaded into the shared memory at a
+ * time.
+ *
+ * @param[in] in_chars Character buffer for the input string column.
+ * @param[in] in_offsets Offset value of each string associated with `in_chars`.
+ * @param[out] out_chars Character buffer for the output string column.
+ * @param[in] out_offsets Offset value of each string associated with `out_chars`.
+ * @param[in] num_strings Number of strings to decode and copy.
  */
-struct url_decode_offsets_updater {
-  device_span<size_type const> const d_esc_positions{};
-  int32_t const first_input_offset = 0;
-
-  /**
-   * @brief Convert input offsets into output offsets
-   *
-   * Each offset is reduced by 2 for every escape sequence that occurs in the entire string column
-   * character data before the offset, as 3 characters are replaced with 1 for each escape.
-   *
-   * @param offset An original offset value from the input string column
-   * @return Adjusted offset value
-   */
-  __device__ int32_t operator()(int32_t offset) const
-  {
-    // determine the number of escape sequences occurring before this offset
-    size_type const* next_esc_pos_ptr =
-      thrust::lower_bound(thrust::seq, d_esc_positions.begin(), d_esc_positions.end(), offset);
-    size_type num_prev_esc = next_esc_pos_ptr - d_esc_positions.data();
-    // every escape that occurs before this one replaces 3 characters with 1
-    return offset - first_input_offset - (num_prev_esc * 2);
+template <int num_warps_per_threadblock, int char_block_size>
+__global__ void url_decode_char_replacer(char const* const in_chars,
+                                         offset_type const* const in_offsets,
```

Comment:
Same comment as per the other kernel. You can pass a `column_device_view` of the input strings here.